azure docs rewrite (#6087)
nextflow-io/nextflow: docs/azure.md

### Managed identities

:::{versionadded} 24.05.0-edge
:::

[Managed identity](https://learn.microsoft.com/en-us/entra/identity/managed-identities-azure-resources/overview) is the recommended authentication method when running Nextflow within Azure. It automatically manages credentials without requiring you to store secrets.

When using a managed identity, an Azure resource can authenticate based on what it is rather than using access keys. For example, Nextflow can submit jobs to Azure Batch and access private storage accounts without additional credentials if it is running on Azure Virtual Machines with a managed identity and the necessary permissions.

There are two types of managed identities:

- System-assigned managed identity
- User-assigned managed identity

**System-assigned managed identity**

A system-assigned identity is tied to a specific Azure resource. To use it:

1. Enable [system-assigned managed identity](https://learn.microsoft.com/en-us/entra/identity/managed-identities-azure-resources/how-to-configure-managed-identities?pivots=qs-configure-portal-windows-vm#system-assigned-managed-identity) on your Azure resource.

> 2. Configure the required role assignments. See [Required Roles](#required-roles) for more information.

Reviewer: ```suggestion
2. Configure the required role assignments. See [Required roles](#required-roles) for more information.
```

    }
    batch {
        accountName = '<YOUR BATCH ACCOUNT NAME>'
        location = '<YOUR BATCH ACCOUNT LOCATION>'
    }
}
```

:::{note}
Environment variable: `AZURE_MANAGED_IDENTITY_SYSTEM=true` can be used instead of the config setting
:::

#### User-Assigned Managed Identity

A user-assigned identity can be shared across multiple Azure resources and its lifecycle is not tied to any specific resource. See [Azure Documentation](https://learn.microsoft.com/en-us/entra/identity/managed-identities-azure-resources/managed-identity-best-practice-recommendations#choosing-system-or-user-assigned-managed-identities) for details.

To use it:

1. [Create a Managed Identity](https://learn.microsoft.com/en-us/entra/identity/managed-identities-azure-resources/how-manage-user-assigned-managed-identities?pivots=identity-mi-methods-azp#create-a-user-assigned-managed-identity)
2. Configure the required role assignments
3. [Assign the identity to your Azure resource](https://learn.microsoft.com/en-us/entra/identity/managed-identities-azure-resources/how-to-configure-managed-identities?pivots=qs-configure-portal-windows-vm#user-assigned-managed-identity)

> 4. Use this configuration:

Reviewer 2: ```suggestion
4. Add this configuration:
```

```groovy
azure {
    managedIdentity {

>         clientId = '<USER ASSIGNED MANAGED IDENTITY CLIENT ID>'

Reviewer 2: ```suggestion
        clientId = '<USER_ASSIGNED_MANAGED_IDENTITY_CLIENT_ID>'
```

    }
    storage {

>         accountName = '<YOUR STORAGE ACCOUNT NAME>'
>     }
>     batch {
>         accountName = '<YOUR BATCH ACCOUNT NAME>'
>         location = '<YOUR BATCH ACCOUNT LOCATION>'

Reviewer 2: ```suggestion
        accountName = '<YOUR STORAGE ACCOUNT NAME>'
    }
    batch {
        accountName = '<BATCH_ACCOUNT_NAME>'
        location = '<BATCH_ACCOUNT_LOCATION>'
```

    }
}
```

:::{note}

> Environment variable: `AZURE_MANAGED_IDENTITY_USER` can be used instead of the clientId setting

Reviewer 2: ```suggestion
The environment variable `AZURE_MANAGED_IDENTITY_USER` can be used instead of the `clientId` setting
```

:::

### Service Principals

:::{versionadded} 22.11.0-edge
:::

[Service Principal](https://learn.microsoft.com/en-us/azure/active-directory/develop/howto-create-service-principal-portal) credentials can be used to access Azure Batch and Storage accounts. Similar to a Managed Identity, a Service Principal is an application identity that can have specific permissions and role-based access. Unlike Managed Identities, Service Principals require a secret key for authentication. This approach allows you to authenticate with Azure resources when operating outside of the Azure environment, such as from your local machine or a non-Azure cloud.

```groovy

> azure {
>     activeDirectory {
>         servicePrincipalId = '<YOUR SERVICE PRINCIPAL CLIENT ID>'
>         servicePrincipalSecret = '<YOUR SERVICE PRINCIPAL CLIENT SECRET>'
>         tenantId = '<YOUR TENANT ID>'
>     }
>     storage {
>         accountName = '<YOUR STORAGE ACCOUNT NAME>'
>     }
>     batch {
>         accountName = '<YOUR BATCH ACCOUNT NAME>'
>         location = '<YOUR BATCH ACCOUNT LOCATION>'
>     }

Reviewer 2: ```suggestion
azure {
    activeDirectory {
        servicePrincipalId = '<SERVICE_PRINCIPAL_CLIENT_ID>'
        servicePrincipalSecret = '<SERVICE_PRINCIPAL_CLIENT_SECRET>'
        tenantId = '<TENANT_ID>'
    }
    storage {
        accountName = '<STORAGE_ACCOUNT_NAME>'
    }
    batch {
        accountName = '<BATCH_ACCOUNT_NAME>'
        location = '<BATCH_ACCOUNT_LOCATION>'
    }
```

}
```

:::{note}
Environment variables can be used instead of config settings:
- `AZURE_CLIENT_ID` - Service principal client ID
- `AZURE_CLIENT_SECRET` - Service principal secret

> - `AZURE_TENANT_ID` - Tenant ID

Reviewer 2: ```suggestion
Environment variables can be used instead of config settings:

- `AZURE_CLIENT_ID`: Service principal client ID.
- `AZURE_CLIENT_SECRET`: Service principal secret.
- `AZURE_TENANT_ID`: Tenant ID.
```

:::

### Access Keys

The basic authentication method using direct access keys. While simple, it's less secure than Managed Identities or Service Principals. This approach requires you to obtain and manage the access keys for your Azure Storage and Batch accounts.

You can find your access keys in the Azure Portal:

> - For Storage accounts: Navigate to your Storage account → "Access keys" in the left menu
> - For Batch accounts: Navigate to your Batch account → "Keys" in the left menu

Reviewer 2: ```suggestion
- For Storage accounts, navigate to your Storage account and select **Access keys** in the left-hand navigation menu.
- For Batch accounts, navigate to your Batch account and select **Keys** in the left-hand navigation menu.
```

> These keys should be kept secure and rotated periodically as they provide full access to your resources. Consider using environment variables rather than hardcoding them in configuration files, especially in shared or version-controlled environments.

Reviewer 2: ```suggestion
Keep access keys secure and rotate them regularly, as they provide full access to your resources. To reduce risk, avoid hardcoding keys in configuration files, use environment variables instead, especially in shared or version-controlled environments.
```

```groovy
azure {
    storage {

>         accountName = '<YOUR STORAGE ACCOUNT NAME>'
>         accountKey = '<YOUR STORAGE ACCOUNT KEY>'

Reviewer 2: ```suggestion
        accountName = '<STORAGE_ACCOUNT_NAME>'
        accountKey = '<STORAGE_ACCOUNT_KEY>'
```

    }
    batch {
        accountName = '<YOUR BATCH ACCOUNT NAME>'
        accountKey = '<YOUR BATCH ACCOUNT KEY>'

>         location = '<YOUR LOCATION>'

Reviewer 2: ```suggestion
        location = '<LOCATION>'
```

    }
}
```

You can also use a Shared Access Token (SAS) instead of an account key:

```groovy

> azure.storage.sasToken = '<YOUR SAS TOKEN>'

Reviewer 2: ```suggestion
azure.storage.sasToken = '<SAS_TOKEN>'
```

```

:::{tip}
When creating a SAS token, ensure you enable these permissions: `Read`, `Write`, `Delete`, `List`, `Add`, `Create` on the resource types `Container` and `Object`.
The value of `sasToken` should be stripped of the leading `?` character.
:::

:::{note}
Environment variables can be used instead of config settings:
- `AZURE_STORAGE_ACCOUNT_NAME` - Storage account name
- `AZURE_STORAGE_ACCOUNT_KEY` - Storage account key
- `AZURE_BATCH_ACCOUNT_NAME` - Batch account name
- `AZURE_BATCH_ACCOUNT_KEY` - Batch account key

> - `AZURE_STORAGE_SAS_TOKEN` - Storage SAS token

Reviewer 2: ```suggestion
Environment variables can be used instead of config settings:

- `AZURE_STORAGE_ACCOUNT_NAME`: Storage account name.
- `AZURE_STORAGE_ACCOUNT_KEY`: Storage account key.
- `AZURE_BATCH_ACCOUNT_NAME`: Batch account name.
- `AZURE_BATCH_ACCOUNT_KEY`: Batch account key.
- `AZURE_STORAGE_SAS_TOKEN`: Storage SAS token.
```

:::

## Azure Blob Storage

Azure Storage provides several services for storing and accessing data in Azure. Nextflow primarily supports Azure Blob Storage for most workloads, with limited support for Azure File Shares for specific use cases. This section describes how to configure and use Azure Storage with Nextflow, focusing on the recommended Blob Storage approach.

> When using Azure Storage with Nextflow, you can access your data using the `az://` URI scheme, which allows transparent access to files stored in Azure Blob Storage. Once the Blob Storage credentials are configured, you can access files in any blob container as if they were local files by prepending the path with `az://` followed by the container name.

Reviewer 2: ```suggestion
When using Azure Storage with Nextflow, you can access your data using the `az://` URI scheme. It allows transparent access to files stored in Azure Blob Storage. Once the Blob Storage credentials are configured, you can access files in any blob container as if they were local files by prepending the path with `az://` followed by the container name.
```

For example:
```groovy
// Access a file in a container
file('az://my-container/path/to/file.txt')
```

The Blob storage account name and key need to be provided in the Nextflow configuration file as shown below:

```groovy
azure {
    storage {

>         accountName = "<YOUR BLOB ACCOUNT NAME>"

Reviewer 2: ```suggestion
        accountName = "<BLOB_ACCOUNT_NAME>"
```

    }
}
```

## Azure Batch

[Azure Batch](https://learn.microsoft.com/en-us/azure/batch/) is a managed computing service that enables large-scale parallel and high-performance computing (HPC) batch jobs in Azure. Nextflow provides a built-in executor for Azure Batch, allowing you to offload your workflows to the cloud for scalable and cost-effective execution.

> Nextflow integrates seamlessly with Azure Batch to:
> - Dynamically create and manage compute pools based on workflow requirements
> - Automatically scale nodes up and down as needed during workflow execution
> - Support both regular and low-priority VMs for cost optimization
> - Handle container execution across distributed compute resources
> - Manage task dependencies and data transfers between Azure Storage and compute nodes

Reviewer 2: ```suggestion
Nextflow integrates seamlessly with Azure Batch to:

- Dynamically create and manage compute pools based on workflow requirements.
- Automatically scale nodes up and down as needed during workflow execution.
- Support both regular and low-priority VMs for cost optimization.
- Handle container execution across distributed compute resources.
- Manage task dependencies and data transfers between Azure Storage and compute nodes.
```

This section describes how to configure and use Azure Batch with Nextflow for efficient cloud-based workflow execution.

For comprehensive information about Azure Batch features and capabilities, refer to the [official Azure Batch documentation](https://learn.microsoft.com/en-us/azure/batch/).

### Overview

> TODO: How it works here.

Reviewer 2: TODO

- pools
- jobs
- tasks
- shared file system using Blob Storage

### Pool Management

:::{warning}
To avoid any extra charges in the Batch account, remember to clean up the Batch pools or use auto scaling.
:::

Nextflow supports two approaches for managing Batch pools:

#### Auto Pools

> When using the autoPoolMode option, Nextflow automatically creates pools of compute nodes appropriate for your pipeline.

Reviewer 2: ```suggestion
When using the `autoPoolMode` option, Nextflow automatically creates pools of compute nodes appropriate for your pipeline.
```

To enable this feature, use one or more Process selectors in your Nextflow configuration to apply the Azure Batch configuration to the subset of processes that you want to offload. For example:

```groovy
process {
    withLabel: bigTask {
        executor = 'azurebatch'
        queue = 'my-batch-pool'
        container = 'my/image:tag'
    }
}

azure {
    storage {
        accountName = '<YOUR STORAGE ACCOUNT NAME>'
        accountKey = '<YOUR STORAGE ACCOUNT KEY>'
    }
    batch {
        location = '<YOUR LOCATION>'
        accountName = '<YOUR BATCH ACCOUNT NAME>'
        accountKey = '<YOUR BATCH ACCOUNT KEY>'
    }
}
```

With the above configuration, processes with the bigTask label will run on Azure Batch, while the remaining processes will run on the local computer.

Then launch the pipeline with the -bucket-dir option to specify an Azure Blob Storage path for the jobs computed with Azure Batch, and optionally, use the -work-dir option to specify the local storage for the jobs computed locally:

```bash

> nextflow run <script> -bucket-dir az://my-container/some/path

Reviewer 2: ```suggestion
nextflow run <SCRIPT> -bucket-dir az://my-container/some/path
```

```

:::{warning}

> - Blob Storage path needs a subdirectory (e.g., `az://container/work` not `az://container`)

Reviewer 2: ```suggestion
- Blob Storage path needs a subdirectory, for example, `az://container/work`, not `az://container`.
```

> - Nextflow handles file transfers between local and cloud
> - With [Fusion](./fusion.md), `-bucket-dir` is optional

Reviewer 2: ```suggestion
- Nextflow handles file transfers between local and cloud.
- With [Fusion](./fusion.md), `-bucket-dir` is optional.
```
